### PR TITLE
fix: fix an issue with wrapping lr_scheduler

### PR DESCRIPTION
**Pull request by hamidzr**
## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/pytorch/lightning/_adapter.py [hunk 1]
```diff
@@ -266,18 +277,20 @@ def lightning_scheduler_dict_to_det(lrs: dict) -> LRScheduler:
...
-                    opt,
-                    self._pls.optimizers,
-                    "Must use an optimizer that is returned by wrap_optimizer()",
-                )
-            return LRScheduler(lrs["scheduler"], step_mode, frequency=lrs["frequency"])
 
-        optimizers = [self._pls.context.wrap_optimizer(opt) for opt in optimizers]
+            wrapped_opt = optimizers_dict[getattr(lrs["scheduler"], "optimizer", None)]
+            if wrapped_opt is None:
+                raise InvalidModelException("missing optimizer for lr_scheduler")
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/lightning/_adapter.py:283]: As it is written, users who hit this message not know how to fix the issue or why they are encountering it.  Please make this message clearer as to what we are doing, why we are doing it, and how the user can work around the issue.

See `wrap_lr_scheduler()` for an example of what I mean:

```python
    raise det.errors.InvalidExperimentException(
        "detected that context.wrap_lr_scheduler() was called with an instance of "
        "torch.optim.lr_scheduer.ReduceLROnPlateau as the lr_scheduler.  This lr "
        "scheduler class does not have the usual step() parameters, and so it can "
        "only be used with step_mode=MANUAL_STEP.\n"
        "\n"
        "For example, if you wanted to step it on every validation step, you might "
        "wrap your lr_scheduler and pass it to a callback like this:\n"
        "\n"
        "class MyLRStepper(PyTorchCallback):\n"
        "    def __init__(self, wrapped_lr_scheduler):\n"
        "        self.wrapped_lr_scheduler = wrapped_lr_scheduler\n"
        "\n"
        "    def on_validation_end(self, metrics):\n"
        '        self.wrapped_lr_scheduler.step(metrics["validation_error"])\n'
    )
```

Comment (rb-determined-ai) [reply harness/determined/pytorch/lightning/_adapter.py:283]: Users will definitely break this.  I was helping a user who was passing a custom LRScheduler without a `step` method just a couple days ago.

File: harness/determined/pytorch/lightning/_adapter.py [hunk 2]
```diff
@@ -266,18 +277,31 @@ def lightning_scheduler_dict_to_det(lrs: dict) -> LRScheduler:
...
+                    "LRSchedulers (or lr_dict).\n"
+                    "- Dictionary, with an ‘optimizer’ key, and (optionally) a ‘lr_scheduler’ key"
+                    "whose value is a single LR scheduler or lr_dict.\n"
+                    "- Tuple of dictionaries as described, with an optional ‘frequency’ key.\n"
                 )
-            return LRScheduler(lrs["scheduler"], step_mode, frequency=lrs["frequency"])
 
-        optimizers = [self._pls.context.wrap_optimizer(opt) for opt in optimizers]
+            # switch the user's unwrapped optimizer with the wrapped version.
+            lrs["scheduler"].optimizer = wrapped_opt
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/lightning/_adapter.py:297]: Can you also emit a check here that the scheduler object actually has an `optimizer` attribute?

Sorry, I didn't understand the check above, I thought that is what I was asking for before.

But the error message above looks fantastic : )

**Review by rb-determined-ai — APPROVED** [review]

**Review by shiyuann — APPROVED** [review]

File: harness/determined/pytorch/lightning/_adapter.py (hunk 2 above)
Comment (hamidzr) [reply harness/determined/pytorch/lightning/_adapter.py:297]: : ) . it seems like the base lrscheduler has a [similar check ](https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L28) but doesn't hurt to have it here as well.

Comment (rb-determined-ai) [reply harness/determined/pytorch/lightning/_adapter.py:297]: Yes, but the issue is that people pass LRSchedulers which are not actually API compatible with the real _LRScheduler at all
